pinot-common/src/main/java/org/apache/pinot/common/utils/config/QueryOptionsUtils.java

```diff
@@ -261,25 +243,63 @@ public static String getOrderByAlgorithm(Map<String, String> queryOptions) {
   @Nullable
   public static Integer getMultiStageLeafLimit(Map<String, String> queryOptions) {
     String maxLeafLimitStr = queryOptions.get(QueryOptionKey.MULTI_STAGE_LEAF_LIMIT);
-    return maxLeafLimitStr != null ? Integer.parseInt(maxLeafLimitStr) : null;
+    return checkedParseInt(QueryOptionKey.MULTI_STAGE_LEAF_LIMIT, maxLeafLimitStr);
   }
 
   @Nullable
   public static Integer getNumGroupsLimit(Map<String, String> queryOptions) {
     String maxNumGroupLimit = queryOptions.get(QueryOptionKey.NUM_GROUPS_LIMIT);
-    return maxNumGroupLimit != null ? Integer.parseInt(maxNumGroupLimit) : null;
+    return checkedParseInt(QueryOptionKey.NUM_GROUPS_LIMIT, maxNumGroupLimit);
   }
 
   @Nullable
   public static Integer getMaxInitialResultHolderCapacity(Map<String, String> queryOptions) {
     String maxInitResultCap = queryOptions.get(QueryOptionKey.MAX_INITIAL_RESULT_HOLDER_CAPACITY);
-    return maxInitResultCap != null ? Integer.parseInt(maxInitResultCap) : null;
+    return checkedParseInt(QueryOptionKey.MAX_INITIAL_RESULT_HOLDER_CAPACITY, maxInitResultCap);
   }
 
   @Nullable
   public static Integer getGroupTrimThreshold(Map<String, String> queryOptions) {
     String groupByTrimThreshold = queryOptions.get(QueryOptionKey.GROUP_TRIM_THRESHOLD);
-    return groupByTrimThreshold != null ? Integer.parseInt(groupByTrimThreshold) : null;
+    return checkedParseInt(QueryOptionKey.GROUP_TRIM_THRESHOLD, groupByTrimThreshold);
+  }
+
+  private static Long checkedParseLong(String optionName, String optionValue, int minValue) {
+    try {
+      if (optionValue != null) {
+        Long value = Long.parseLong(optionValue);
+        if (value < minValue) {
+          throw new IllegalStateException(
+              Strings.lenientFormat("%s must be a number between %s and 2^63-1, got: %s", optionName, minValue,
+                  optionValue));
+        }
+        return value;
+      } else {
+        return null;
+      }
+    } catch (NumberFormatException nfe) {
+      throw new IllegalStateException(
+          Strings.lenientFormat("%s must be a number between %s and 2^63-1, got: %s", optionName, minValue,
+              optionValue));
+    }
+  }
+
+  private static Integer checkedParseInt(String optionName, String optionValue) {
+    try {
+      if (optionValue != null) {
+        int value = Integer.parseInt(optionValue);
+        if (value < 0) {
+          throw new IllegalStateException(
+              Strings.lenientFormat("%s must be a number between 0 and 2^31-1, got: %s", optionName, optionValue));
```

Comment:
Any reason to use `Strings.lenientFormat` over `String.format` here?

Reply:
That's a leftover from earlier partially-applied check with Precondition class. Same reason I used IllegalStateException.

Reply:
Yeah, I think `IllegalArgumentException` might be a better fit?

Reply:
Replaced with String.format().